src/tests/end2end/CopyTests.cpp

```diff
@@ -0,0 +1,418 @@
+                .SetAllowedUsage(nxt::TextureUsageBit::TransferDst | nxt::TextureUsageBit::TransferSrc)
+                .GetResult();
+
+            nxt::CommandBuffer commands = device.CreateCommandBufferBuilder()
+                .TransitionTextureUsage(texture, nxt::TextureUsageBit::TransferDst)
+                .CopyBufferToTexture(buffer, offset, rowPitch, texture, 0, 0, 0, width, height, 1, 0)
+                .GetResult();
+
+            queue.Submit(1, &commands);
+
+            EXPECT_TEXTURE_RGBA8_EQ(expected.data(), texture, 0, 0, width, height, 0) << "Failed to copy to " << width << "x" << height << " texture with buffer offset " << offset << std::endl;
+        }
+
+        // Copy from a buffer with double the needed row pitch
+        void TestRowPitch(uint32_t width, uint32_t height) {
+            uint32_t rowPitch = Align(width * kBytesPerTexel, kTextureRowPitchAlignment);
+            uint32_t bufferRowPitch = rowPitch * 2;
+            uint32_t bufferSize = bufferRowPitch * (height - 1) + width * kBytesPerTexel;
+
+            std::vector<RGBA8> data(bufferSize / kBytesPerTexel);
+            std::vector<RGBA8> expected(width * height);
+
+            FillTextureData(width, height, bufferRowPitch / kBytesPerTexel, data.data());
+            PackTextureData(data.data(), width, height, bufferRowPitch / kBytesPerTexel, expected.data(), width);
+
+            nxt::Buffer buffer = device.CreateBufferBuilder()
+                .SetAllowedUsage(nxt::BufferUsageBit::TransferDst | nxt::BufferUsageBit::TransferSrc)
+                .SetInitialUsage(nxt::BufferUsageBit::TransferDst)
+                .SetSize(bufferSize)
+                .GetResult();
+
+            buffer.SetSubData(0, static_cast<uint32_t>(data.size()), reinterpret_cast<const uint32_t*>(data.data()));
+            buffer.FreezeUsage(nxt::BufferUsageBit::TransferSrc);
+
+            nxt::Texture texture = device.CreateTextureBuilder()
+                .SetDimension(nxt::TextureDimension::e2D)
+                .SetExtent(width, height, 1)
+                .SetFormat(nxt::TextureFormat::R8G8B8A8Unorm)
+                .SetMipLevels(1)
+                .SetAllowedUsage(nxt::TextureUsageBit::TransferDst | nxt::TextureUsageBit::TransferSrc)
+                .GetResult();
+
+            nxt::CommandBuffer commands = device.CreateCommandBufferBuilder()
+                .TransitionTextureUsage(texture, nxt::TextureUsageBit::TransferDst)
+                .CopyBufferToTexture(buffer, 0, bufferRowPitch, texture, 0, 0, 0, width, height, 1, 0)
+                .GetResult();
+
+            queue.Submit(1, &commands);
+
+            EXPECT_TEXTURE_RGBA8_EQ(expected.data(), texture, 0, 0, width, height, 0) << "Failed to copy to " << width << "x" << height << " texture with buffer row pitch " << bufferRowPitch << std::endl;
+        }
+};
+
+// Test that creating and reading back an entire texture with 256-byte aligned dimensions works
+TEST_P(CopyTests_T2B, FullTextureAligned) {
```

Comment:
T2B versions of the test are testing the texture-related parameters and B2T are testing the buffer-related ones. We basically have half-coverage of the T<->B copies. Can we make both T->B and B->T test both?